### ISSUE TITLE
VMs with TrueNAS iSCSI disks are not compatible with Proxmox HA 

### ISSUE THREAD

**Issue by antoninchadima** [TheGrandWazoo/freenas-proxmox#179]
Everything is fine without HA.

With HA the VM doesn't start.
```
task started by HA resource agent
TASK ERROR: freenas: unknown iscsi provider. Available [comstar, istgt, iet, LIO] at /usr/share/perl5/PVE/Storage/ZFSPlugin.pm line 35.
```

**Comment by TheGrandWazoo** [comment]
Did you install the plugin on each node in the cluster?

**Comment by antoninchadima** [comment]
yes it is installed on all 12 nodes
i can even live migrate a vm from one node to another

but a VM with TrueNAS iSCSI disks is not compatible with being added as a resource in Proxmox HA

![Screenshot from 2024-03-23 07-27-39](https://github.com/TheGrandWazoo/freenas-proxmox/assets/3483314/18a563cb-0a5a-4643-a40d-3be91c2fb15f)

![Screenshot from 2024-03-23 07-32-36](https://github.com/TheGrandWazoo/freenas-proxmox/assets/3483314/c0f3a6b3-a8ca-4142-b69e-05c185cf3d91)

**Comment by TheGrandWazoo** [comment]
I have HA running on my cluster without issues. Are you sure the target host has the Plugin installed? Due to the fact that the error is:
`TASK ERROR: freenas: unknown iscsi provider. Available [comstar, istgt, iet, LIO] at /usr/share/perl5/PVE/Storage/ZFSPlugin.pm line 35.`
And the `[comstar, istgt, iet, LIO]` does NOT have the FreeNAS leads me to believe that node did not get the Plugin installed...

Snippet from the patch for ZFSPlugin.pm showing the original and the patch.
```
-    die "$provider: unknown iscsi provider. Available [comstar, istgt, iet, LIO]";
+    die "$provider: unknown iscsi provider. Available [comstar, freenas, istgt, iet, LIO]";
```

**Comment by antoninchadima** [comment]
thank you again for your time

the plugin is surely installed on every node
the storage is visible on every node in the server view on the right side
and in the Datacenter (cluster1) Search overview:
![image](https://github.com/TheGrandWazoo/freenas-proxmox/assets/3483314/f0e87f39-cb67-4705-aaf0-d71425a0d571)

The HA is defined as follows (there are no other nodes):
![image](https://github.com/TheGrandWazoo/freenas-proxmox/assets/3483314/6fafed5c-b370-4809-b143-fde3cc29844c)

the storage is defined for all nodes no exception:
![image](https://github.com/TheGrandWazoo/freenas-proxmox/assets/3483314/244637c8-c6f3-438e-aa39-8e829f5bf047)


Maybe i can try to reinstall the plugin on each node again (if it is possible)
and tell you the results

**Comment by antoninchadima** [comment]
did `apt --reinstall install freenas-proxmox` on every node
and got the same message with no errors:
```
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
0 upgraded, 0 newly installed, 1 reinstalled, 0 to remove and 28 not upgraded.
Need to get 2,924 B of archives.
After this operation, 0 B of additional disk space will be used.
Get:1 https://dl.cloudsmith.io/public/ksatechnologies/truenas-proxmox/deb/debian any-version/main amd64 freenas-proxmox all 2.2.0-1 [2,924 B]
Fetched 2,924 B in 0s (32.2 kB/s)           
(Reading database ... 66952 files and directories currently installed.)
Preparing to unpack .../freenas-proxmox_2.2.0-1_all.deb ...
Unpacking freenas-proxmox (2.2.0-1) over (2.2.0-1) ...
Reinstalling original Proxmox VE files so upgrade can be performed
Setting up freenas-proxmox (2.2.0-1) ...
Already up to date.
Configuring freenas-proxmox 2.2.0-1
Restarting pvedaemon...
Restarting pveproxy...
Restarting pvestatd...
Restarting pvescheduler...
```
no signing key and repository issues, on all nodes fine

Maybe a nasty proxmox issue?
The HA existed before adding the iscsi storage...
so i can try to delete the HA group disable HA
and enable it again (after having this plugin installed)
and tell you the results

**Comment by antoninchadima** [comment]
this doesn't help
 i can even reproduce it in a clean fresh installed proxmox 3 node cluster

i can create a VM with this iscsi storage on every node
i can even manually migrate it to every node

The HA Settings in Options is set to Shutdown Policy=migrate
Cluster Resource Scheduling: HA Scheduling= Static Load, Rebalance on Start=true

i create a HA group with all the nodes included (restricted=true, nofailback=true)
i add the VM to this HA group  (max-started=1, max-relocate=1, requested state=started)
everything goes fine up to this point

aftewards i stop the VM

HA 122 - Stop:
```
Requesting HA stop for VM 122
TASK OK
```
VM 122 - Shutdown:
```
task started by HA resource agent
VM still running - terminating now with SIGTERM
TASK OK
```

everything is still OK

and finally i try to start the VM

HA 122 - Start:
```
Requesting HA start for VM 122
TASK OK
```
VM 122 - Migrate:
```
task started by HA resource agent
2024-03-25 09:35:27 use dedicated network address for sending migration traffic (172.16.5.40)
2024-03-25 09:35:28 starting migration of VM 122 to node 'pve10' (172.16.5.40)
2024-03-25 09:35:28 migration finished successfully (duration 00:00:01)
TASK OK
```
VM 122 - Start:
```
task started by HA resource agent
TASK ERROR: freenas: unknown iscsi provider. Available [comstar, istgt, iet, LIO] at /usr/share/perl5/PVE/Storage/ZFSPlugin.pm line 35
```

and the VM with iscsi storage plugin doesn't start and is in HA error state
the same for a clean new Proxmox install, 3 nodes, no other VM
and i don't have any problem with nfs and ceph with HA on the same cluster

**Comment by antoninchadima** [comment]
I have checked every single /usr/share/perl5/PVE/Storage/ZFSPlugin.pm
if patched and contains `die "$provider: unknown iscsi provider. Available [comstar, freenas, istgt, iet, LIO]";`
and i didnt find a single one with only: `die "$provider: unknown iscsi provider. Available [comstar, istgt, iet, LIO]";`

mystery
maybe some cache? will restart the nodes

**Comment by antoninchadima** [comment]
Ok a complete restart of all proxmox nodes helped, even on a completely new installation.

thank you very much and you can close the ticket.
maybe this should be mentioned in the readme.

**Comment by antoninchadima** [comment]
Ok there is no more the HA issue, but:
```
task started by HA resource agent
iscsiadm: No session found.
iscsiadm: No session found.
iscsiadm: No session found.
iscsiadm: No session found.
iscsiadm: No session found.
iscsiadm: No session found.
/dev/rbd4
TASK OK
```

**Comment by mir07** [comment]
I think I have a solution for this issue. The problem is when the hooks for the freenas plugin kicks in an patches the upstream files it runs these systemd commands:

- systemctl restart pvedaemon.service
- systemctl restart pvestatd.service
- systemctl restart pveproxy.service

But the problem is that ha-manager runs with the original ZFSPlugin.pm so to update ha-manager the following systemd sequence should have been run:

- systemctl restart pvedaemon.service
- systemctl restart pve-ha-crm.service
- systemctl restart pve-ha-lrm.service
- systemctl restart pvestatd.service
- systemctl restart pveproxy.service

**Comment by antoninchadima** [comment]
yes, that's it!

Thank you very much.

**Comment by TheGrandWazoo** [comment]
Reopen. Need to provide a fix in the installer.

**Comment by deamonkai** [comment]
This is still present in **v3** — the fix discussed here (also restart `pve-ha-crm`/`pve-ha-lrm`, per @mir07) never landed.

The current v3 `.deb` postinst `restart_pve_services()` restarts only:

```
pvedaemon restart
pveproxy restart
```

— not `pvestatd`, `pve-ha-lrm`, or `pve-ha-crm`. So after a fresh `apt install truenas-proxmox` (v3 track), the HA daemons keep their pre-install `PVE::Storage` state, and HA-managed VMs on `truenas:` storage fail to start:

```
pve-ha-lrm: storage 'TrueNAS' does not exist
pve-ha-lrm: unable to start service vm:<vmid>
```

(the v3 variant of the v2 "unknown iscsi provider" error above — same root cause.)

Repro: PVE 9.2.4, `truenas-proxmox 3.2.3-1`, TrueNAS SCALE. Manual `qm start`, `pvesm status`, and non-HA VMs all work; only HA-initiated starts fail. Restarting `pve-ha-lrm`/`pve-ha-crm` on each node resolves it permanently — no guest disruption (HA services briefly `freeze`, then return to `started`) — confirming @mir07's diagnosis.

Suggested fix: have `restart_pve_services()` (and the README) restart the full set on every node:

```
systemctl restart pvedaemon pveproxy pvestatd pve-ha-lrm pve-ha-crm
```

Would you consider reopening?

**Comment by noberasco** [comment]
I just got hit by the same issue.
It is not a problem with the freenas-proxmox, rather it is a documentation problem.
The whole README.md is rather out of date: the compatibility matrix is correct, the instructions on how to install the plugin on proxmox are OK _except for the restart services issue_, but the documentation on how to actually configure and start using the feature are outdated:
- on TrueNAS side, I have scale 25.04 and the instructions provided point to configuration screens that do not even exist any more. I had to guess how to do it, and it was non-obvious. The option to create a token is in a different section. Then I had to configure a complete iscsi share on TrueNAS following the wizard. Only after this I was able to successfully configure the new storage on Proxmox. Finally, after I had Proxmox configure the first share of its own, I deleted the one I had created manually on TrueNAS, keeping the rest of the setup that was needed
- on ProxMox side, the instructions are completely out of kilter. You don't add a new 'ZFS over iSCSI', you hadd a new 'TrueNAS (ZFS/ISCSI)' and the fields you have to fill-in are different.

**Comment by TheGrandWazoo** [comment]
Reopening — confirmed still present in v3 (postinst `restart_pve_services()` at packaging/DEBIAN/postinst:41-46 only restarts `pvedaemon` and `pveproxy`). Per @mir07's original diagnosis and @deamonkai's v3 confirmation, the fix is to also restart the HA daemons on install/upgrade so they pick up the patched `PVE::Storage` state:

```
systemctl restart pvedaemon pveproxy pvestatd pve-ha-lrm pve-ha-crm
```

Will track the postinst fix under this issue. Filed a separate issue for the outdated README/SCALE 25.04 docs @noberasco raised, since that's an unrelated documentation gap.

**Comment by TheGrandWazoo** [comment]
Docs issue filed: #286

**Comment by TheGrandWazoo** [comment]
Fix shipped in v3.2.4 (currently on the testing channel — commit d60adfa/eadf692), with a regression test (tests/test-restart-services.sh, runs in CI) and a lab runbook for the real-cluster check (.claude/cos/runbooks/RUNBOOK-001-ha-restart-validation.md).

Tested on pve01-hq, pve02-hq, pve03-hq (PVE 9.2.3/8.4.19):
- `apt install` the 3.2.4~beta build → postinst log shows all 5 services restarted (`pvedaemon`, `pveproxy`, `pvestatd`, `pve-ha-lrm`, `pve-ha-crm`), all confirmed `active` afterward.
- Full repro on pve01-hq: added a stopped TrueNAS-backed VM (disk on `TrueNAS-Scale01-Tank01`) to a temporary HA group, `ha-manager set --state stopped` then `--state started`. Task log: `task started by HA resource agent` ... `TASK OK` — no `storage does not exist` / `unknown iscsi provider` errors. Cleaned up the temporary HA group/resource afterward.

Note: `pve-ha-lrm`/`pve-ha-crm` don't support a `restart` subcommand (only start/stop/status) — they're restarted via `systemctl restart` instead of the `<daemon> restart` form used for the others.

Leaving open until this promotes to stable and gets a final confirm from the UI side, per usual process.

**Comment by TheGrandWazoo** [comment]
(Auto-closed by the `Fixes #179` commit trailer when d60adfa landed on `release/3.x` — leaving the full closing summary here.)

Shipped in **v3.2.4-1** (stable), tag [v3.2.4](https://github.com/TheGrandWazoo/freenas-proxmox/releases/tag/v3.2.4), commits:
- 55c2663 / d60adfa — restart pvestatd + pve-ha-lrm/pve-ha-crm (via systemctl) on install/remove; switched restarts from `cmd && log ...` to bare statements so a genuine failure aborts instead of being silently swallowed under `set -e`
- eadf692 — CI fix (executable bit on new test stubs)

Tested end-to-end on pve01-hq/pve02-hq/pve03-hq, including a full HA `stopped`→`started` cycle on a TrueNAS-backed VM (see earlier comment for the task log). Now on the stable apt dist.

Regression coverage: `tests/test-restart-services.sh` (runs in CI on every push) + `.claude/cos/runbooks/RUNBOOK-001-ha-restart-validation.md` for manual real-cluster checks.

Filed #287 to port the same fix to `truenas-proxmox-multipath`, which has the same gap and wasn't part of this fix.

Thanks @mir07 for the original root-cause diagnosis and @deamonkai for confirming it was still present in v3.
